pkg/envoy/registry/services.go

```diff
@@ -27,34 +34,162 @@ func (e ExplicitProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service
 type KubeProxyServiceMapper struct {
-	KubeController k8s.Controller
+	kubeController k8s.Controller
+	kubeEvents     chan interface{}
+	servicesForCN  map[certificate.CommonName][]service.MeshService
+	cacheLock      sync.RWMutex
 }
 
-// ListProxyServices maps an Envoy instance to a number of Kubernetes services.
-func (k *KubeProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service.MeshService, error) {
-	cn := p.GetCertificateCommonName()
+// NewKubeProxyServiceMapper initializes a KubeProxyServiceMapper with an empty cache.
+func NewKubeProxyServiceMapper(controller k8s.Controller) *KubeProxyServiceMapper {
+	return &KubeProxyServiceMapper{
+		kubeController: controller,
+		servicesForCN:  make(map[certificate.CommonName][]service.MeshService),
+		kubeEvents: events.GetPubSubInstance().Subscribe(
+			announcements.PodAdded,
+			announcements.PodUpdated,
+			announcements.PodDeleted,
+			announcements.ServiceAdded,
+			announcements.ServiceUpdated,
+			announcements.ServiceDeleted,
+		),
+	}
+}
 
-	pod, err := envoy.GetPodFromCertificate(cn, k.KubeController)
-	if err != nil {
-		return nil, err
+// Run starts updating the proxy-to-services cache based on k8s notifications.
+func (k *KubeProxyServiceMapper) Run(stop <-chan struct{}) {
+	for {
+		select {
+		case <-stop:
+			return
+		case ev := <-k.kubeEvents:
+			event, ok := ev.(events.PubSubMessage)
+			if !ok {
+				log.Error().Msgf("ignoring unexpected pubsub message type: %T %v", ev, ev)
+				continue
+			}
+			switch event.AnnouncementType {
+			case announcements.PodAdded, announcements.PodUpdated:
+				pod := event.NewObj.(*v1.Pod)
+				k.handlePodUpdate(pod)
+			case announcements.PodDeleted:
+				pod := event.OldObj.(*v1.Pod)
+				k.handlePodDelete(pod)
+			case announcements.ServiceAdded, announcements.ServiceUpdated:
+				svc := event.NewObj.(*v1.Service)
+				k.handleServiceUpdate(svc)
+			case announcements.ServiceDeleted:
+				svc := event.OldObj.(*v1.Service)
+				k.handleServiceDelete(svc)
+			}
+			events.GetPubSubInstance().Publish(events.PubSubMessage{
+				AnnouncementType: announcements.ScheduleProxyBroadcast,
+			})
+		}
 	}
+}
 
-	services, err := listServicesForPod(pod, k.KubeController)
+func (k *KubeProxyServiceMapper) handlePodUpdate(pod *v1.Pod) {
+	if pod == nil {
+		return
+	}
+	cn, err := getCertCommonNameForPod(*pod)
 	if err != nil {
-		return nil, err
+		log.Error().Err(err).Msgf("ignoring updated pod %s/%s", pod.Namespace, pod.Name)
+		return
 	}
 
-	if len(services) == 0 {
-		return nil, nil
-	}
+	services := listServicesForPod(pod, k.kubeController)
 
 	meshServices := kubernetesServicesToMeshServices(services)
 
 	servicesForPod := strings.Join(listServiceNames(meshServices), ",")
 	log.Trace().Msgf("Services associated with Pod with UID=%s Name=%s/%s: %+v",
 		pod.ObjectMeta.UID, pod.Namespace, pod.Name, servicesForPod)
 
-	return meshServices, nil
+	k.cacheLock.Lock()
```

Comment:
QoL suggestion, not sure everybody likes this as it comes down the more or less the same number of lines of code:
```
func (k *KubeProxyServiceMapper) withLock(f func()) {
  k.cacheLock.Lock()
  defer k.cacheLock.Unock()
  f()
}
```

then 
```
   k.withLock(func() {
     k.servicesForCN[cn] = meshServices
   })
```